Improve exception handling in .NET isolated

andystaples opened the pull request:
<!-- Start the PR description with some context for the change. -->
Improves exception handling in .NET isolated by serializing the exception details in the worker middleware, then deserializing them back in host middleware. 
This may introduce breaking changes to the way that the Functions Host logs exceptions from Durable activities. 
Also adds testing for error handling, activity retry, and activity timeout scenarios in dotnet-isolated

Resolves #2711 

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)
* [x] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
    * [ ] Otherwise: That work is being tracked here: #issue_or_pr_in_each_sdk
* [x] My changes **do not** change the version of the WebJobs.Extensions.DurableTask package
    * [ ] Otherwise: major or minor version updates are reflected in `/src/Worker.Extensions.DurableTask/AssemblyInfo.cs`
* [x] My changes **do not** add EventIds to our EventSource logs
    * [ ] Otherwise: Ensure the EventIds are within the supported range in our existing Windows infrastructure. You may validate this with a deployed app's telemetry. You may also extend the range by completing a PR such as [this one](https://msazure.visualstudio.com/One/_git/AAPT-Antares-Websites/pullrequest/7463263?_a=files).
* [ ] My changes **should** be added to **v2.x** branch.
    * [x] Otherwise: This change applies exclusively to WebJobs.Extensions.DurableTask v3.x. It will be retained only in the `dev` and `main` branches and will not be merged into the `v2.x` branch.

andystaples committed:
Add error handling and timeout tests

andystaples committed:
Improve error handling in dotnet-isolated
- Serialize exception details in worker middleware
- Attempt to deserialize in host middleware
- Ensure inner exceptions + complex messages are preserved

andystaples committed:
Fix another logging race condition

cgillum commented on test/e2e/Tests/Tests/ErrorHandlingTests.cs:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    [Fact]
+    public async Task OrchestratorWithCaughtActivityException_ShouldSucced()
```
```suggestion
    public async Task OrchestratorWithCaughtActivityException_ShouldSucceed()
```

cgillum replied:
Looks like the same typo is in multiple methods.

andystaples replied:
Fixed

cgillum commented on test/e2e/Tests/Tests/ErrorHandlingTests.cs:
```diff
@@ -0,0 +1,99 @@
+
+        var orchestrationDetails = await DurableHelpers.GetRunningOrchestrationDetailsAsync(statusQueryGetUri);
+        Assert.StartsWith("Microsoft.DurableTask.TaskFailedException", orchestrationDetails.Output);
+        Assert.Contains("This activity failed", orchestrationDetails.Output);
```
Is there anything more we can validate here? For example, the type of the inner exception?

andystaples replied:
Added code to double check the inner exception type in the test functionapp code, and am also validating it in the log output in tests.

cgillum commented on test/e2e/Tests/Tests/ErrorHandlingTests.cs:
```diff
@@ -0,0 +1,99 @@
+    public async Task OrchestratorWithUncaughtActivityException_ShouldFail()
+    {
+        using HttpResponseMessage response = await HttpHelpers.InvokeHttpTrigger("RethrowActivityException_HttpStart", "");
+        string actualMessage = await response.Content.ReadAsStringAsync();
```
Looks like the `actualMessage` variable isn't being used in any of these tests.

andystaples replied:
Fixed

cgillum commented on test/e2e/Tests/Tests/TimeoutTests.cs:
```diff
@@ -0,0 +1,39 @@
+    public async Task TimeoutFunction_ShouldTimeoutWhenAppropriate(int timeoutSeconds, string expectedOutput)
+    {
+        using HttpResponseMessage response = await HttpHelpers.InvokeHttpTrigger("TimeoutOrchestrator_HttpStart", $"?timeoutSeconds={timeoutSeconds}");
+        string actualMessage = await response.Content.ReadAsStringAsync();
```
`actualMessage` is not used here either.

andystaples replied:
Fixed

andystaples committed:
Only wrap activity functions

cgillum commented on src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:
```diff
@@ -666,6 +687,27 @@ private static bool TryGetRpcExceptionFields(
+        {
+            try
+            {
+                string serializedMessage = exception.Split('\n')[0];
```
Two things:
1. `string.Split` can be a bit expensive. It's faster and less memory intense to do `IndexOf` and `Substring`.
2. Let's not attempt to parse the message unless we have some clue that it's JSON to avoid parsing things that obviously aren't JSON.

Suggestion:
```csharp
if (exception[0] != '{') // expecting JSON
{
    details = null;
    return false;
}

int newlineIndex = input.IndexOf('\n');
string failureDetailsJson = newlineIndex < 0 ? exception : exception.Substring(0, newlineIndex).Trim();
P.TaskFailureDetails? taskFailureDetails = JsonConvert.DeserializeObject<P.TaskFailureDetails>(failureDetailsJson);
// ...
```

andystaples replied:
Implemented

cgillum commented on src/Worker.Extensions.DurableTask/Exceptions/DurableSerializationException.cs:
```diff
@@ -0,0 +1,42 @@
+
+    public override string? StackTrace => this.FromException.StackTrace;
+
+    private static TaskFailureDetails? ExceptionToTaskFailureDetailsRecursive(Exception? fromException)
```
Do we have a shared version of this we can use? Or is it in some inaccessible place, so we have to copy it? I'm a tiny bit worried about keeping multiple copies up-to-date if we decide to add richer information to `TaskFailureDetails` (which is very likely).

andystaples replied:
Moved to a new TaskFailureDetailsConverter

cgillum commented on src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs:
```diff
@@ -94,4 +103,33 @@ static async Task RunEntityAsync(
+            await next(functionContext);
+            return;
+        }
+        catch (Exception ex)
+        {
+            throw new DurableSerializationException(ex);
+        }
```
Would it be beneficial to catch and do this for orchestration and entity triggers as well?

andystaples replied:
I will suggest that no, for now, no change is needed for these function types. 

Orchestrations never attempt to parse exception details returned from the host into FailureDetails. 

Entitites do, however, since the entity data is returned through the TaskEntityDispatcher, error message is already parsed by the worker and returned in an intelligent way. 

There is some difference in end result for entities. Here is one example: activity functions that throw exceptions have and continue to throw the same exception type in the orchestration. By contrast, entities throw an EntityOperationFailedException which has a TaskFailureDetails property containing the outermost user code exception. 

Additionally, after the outermost exception thrown by the user code, no inner exceptions are returned when an entity throws. This appears to be deliberate, as inner user exception details were stripped in both the webjobs and worker extensions. We can re-evaluate this decision, but I believe it to be beyond the scope of this PR.

andystaples committed:
PR suggestions round 1

bachuv commented on src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:
```diff
@@ -666,6 +687,34 @@ private static bool TryGetRpcExceptionFields(
+            }
+            catch (Exception)
+            {
+                // Apparently the exception message was not serialized by the worker middleware, continue
```
Do we want to log a message saying what's in this comment? Will this be useful for debugging issues in the future?

andystaples replied:
Done - the actual warning is logged using TraceHelper.ExtensionWarningEvent in CallActivityAsync based on whether the GetFailureDetails call reports successful deserialization of the new message. 
If there is a better way to log this, please, let me know

andystaples replied:
Oh also, a consideration: GetFailureDetails is used for all languages that use OutOfProcMiddleware, so I believe Java/others will now log this warning on every activity failure, as the worker extensions for these languages do not know to serialize the exception details in this way. 
We may eventually want to update the other extensions in a similar way - happy to remove/comment this warning until we expect all languages to support exception serialization, or I can try to check the language and only warn for dotnet, or leave the warning in as-is, if we think that the warning will not be distracting for other languages.

andystaples replied:
Update - added a check for dotnet isolated

bachuv replied:
> as the worker extensions for these languages do not know to serialize the exception details in this way.

Can you clarify what you mean here?

andystaples replied:
As in, the Java worker doesn't have the worker-side logic implemented in this PR - to wrap the exception message in Json so that the WebJobs extension can parse. I've added a task in Java to look at this here https://github.com/microsoft/durabletask-java/issues/205

andystaples committed:
Update tests
- Add tests for entity error handling
- Revert to synchronous worker middleware
- Log a warning if exception message not serialized

andystaples committed:
Only log serialization warning in dotnet-isolated

cgillum commented on src/Worker.Extensions.DurableTask/Exceptions/DurableSerializationException.cs:
```diff
@@ -0,0 +1,26 @@
+
+internal class DurableSerializationException : Exception
+{
+    private Exception FromException;
```
Fields should be lowercased and, whenever possible, should be `readonly`.
```suggestion
    private readonly Exception fromException;
```

andystaples replied:
Done

cgillum commented on test/e2e/Apps/BasicDotNetIsolated/ActivityErrorHandling.cs:
```diff
@@ -0,0 +1,163 @@
+        [DurableClient] DurableTaskClient client,
+        FunctionContext executionContext)
+    {
+        ILogger logger = executionContext.GetLogger("RethrowActivityException_HttpStart");
```
Should this be:
```suggestion
        ILogger logger = executionContext.GetLogger("CatchHttpStart");
```

andystaples replied:
Fixed

cgillum commented on test/e2e/Apps/BasicDotNetIsolated/ActivityErrorHandling.cs:
```diff
@@ -0,0 +1,163 @@
+
+public static class ActivityErrorHandling
+{
+    private static ConcurrentDictionary<string, int> retryCount = new ConcurrentDictionary<string, int>();
```
nit: consider naming this `globalRetryCount` to make it clear that it's a global (static) value.

andystaples replied:
Done

cgillum commented on test/e2e/Apps/BasicDotNetIsolated/TimeoutOrchestration.cs:
```diff
@@ -0,0 +1,70 @@
+        for (int i = 0; i < 5; i++) 
+        {
+            Thread.Sleep(1000);
+        }
```
What's the purpose of having a sleep in a for-loop vs. just doing `Thread.Sleep(5000)`?

andystaples replied:
Carryover from testing, updated

cgillum reviewed (approved):
LGTM - just adding a few more comments/questions.

andystaples committed:
More PR suggestions

andystaples committed:
Merge branch 'dev' into andystaples/add-error-handling-tests
